### ISSUE TITLE
The create twin-hint matches names literally, so it missed 4 of 17 duplicates in #501's triage

### ISSUE THREAD

**Issue by gregoryfoster** [CannObserv/power-map#533]
Part of #490. Found during #501's triage of the 2026-09-16 diff (run `2026-09-16T093332Z`); the blocking producer-side cause is CannObserv/usa-wa#378.

## What happened

`_create_hints` is the one signal standing between an operator and a duplicate person: for each `create` it names the PM rows already carrying the asserted value, so a twin reads as a twin rather than a gap. It is built on `value_matches`, which compares **literally**:

```sql
SELECT <column> AS value, <parent> AS parent
  FROM <table> WHERE <column> = ANY($1::text[])
```

In the triage, all 17 `desired_people` creates turned out to be duplicates of live, already-anchored PM people. **The hint flagged 13.** The four it missed differ from what PM holds only in formatting:

| producer publishes | PM already holds |
|---|---|
| `E. G. “Pat” Patterson` | `E. G. "Pat" Patterson` — curly vs straight quotes |
| `Gerald L. “Jerry” Saling` | `Gerald "Jerry" Saling` — a middle initial |
| `Stanley C. Johnson` | `Stanley Johnson` — a middle initial |
| `Myron “Mike” Kreidler (On leave of absence for military duty Jan. 8, 1991 to April 18, 1991)` | `Mike Kreidler` — a note baked into `name_full` |

Each of those four presented to the operator as **"hint: none — PM holds no row matching the asserted name"**, which is not merely unhelpful: it is affirmative evidence for creating the row. An operator working the diff at face value would have minted four duplicates of people with live role assignments.

## Why it matters beyond this run

The 13 that did match were caught by luck of string equality, not by the check doing its job. The producer controls the name text, so any punctuation change upstream silently converts a caught twin into a missed one — the signal degrades without anything failing.

## Suggested direction

Match on a normalised form rather than the raw string — casefold, normalise Unicode quotes and dashes, collapse whitespace, and ideally compare on a name-part basis so a middle initial does not defeat it. The machinery is not new: duplicate detection already exists (#474 and its spawn #476–#479) and entity search (#316) does its own normalisation. This is about routing the create hint through comparable normalisation rather than inventing one.

Two design points worth settling first:

- **The hint is advisory and must stay cheap.** It runs per create, per run. A fuzzy match that returns half the table is worse than a literal one; a normalised-equality match is probably the right ceiling, with anything softer left to the duplicate audit.
- **Recall over precision, but say which is which.** A hint that names a probable twin costs an operator one look. A hint that stays silent costs a duplicate person. If normalised matching is added, it is worth marking in the hint whether the match was exact or normalised, so the triage record says what was actually compared.

## Acceptance

- The four cases above produce a hint against the PM person they duplicate.
- The 13 exact matches keep theirs.
- A test fixture covering each shape — curly quote, middle initial, trailing parenthetical, case — so this cannot regress quietly.

**Comment by gregoryfoster** [comment]
Producer side here. Two updates, one of which changes a row in your table.

## The Kreidler line is stale

`Myron “Mike” Kreidler (On leave of absence for military duty Jan. 8, 1991 to April 18, 1991)` no longer publishes. CannObserv/usa-wa#378 shipped a screen that removes parentheticals asserting *facts about service* while leaving parentheticals that are *name content* (marital print forms, legal-name glosses), plus a build gate at zero. As of `persons` `v20260917T214318Z-8ec0ea` that row reads:

```
Myron “Mike” Kreidler
```

Still not a literal match for your `Mike Kreidler`, so the finding stands — but the shape it misses on is now "a different given name", not "a note baked into `name_full`".

## The 17 resolve by tombstone, so this is no longer cutover-blocking

`person_crosswalk` `v20260917T214318Z-8ec0ea` carries `merged_into` for all 17 (plus Heck). On rebuild they should present as merges against rows you already hold, not as creates needing a hint. #533's value is the *next* duplicate, not these.

Worth knowing for that next one: **four of those 17 arrive as name updates on anchored rows** — the same four your hint missed. Details on #501.

## A producer-side counterpart now exists, with a different blind spot

usa-wa shipped `persons_distinct_identities`: live entities grouped by the roster↔WSL matcher's own `identity_fold`, un-allowlisted collisions **fail `dbt build`**. Verified zero on the live corpus, red on an injected duplicate.

The overlap with your hint is the useful part, because the two disagree:

| shape | your literal hint | our fold |
|---|---|---|
| curly vs straight quotes | missed | **caught** — the fold normalises them |
| honorifics, quoted nicknames | missed | caught |
| middle initial one side only | missed | missed |
| different given name (Myron/Mike) | missed | missed |

So the normalisation you propose is the right direction and would converge our two screens on the first two rows. **The last two rows are what neither side catches**, and they are not cheap to close: we measured the obvious loosening — first given token plus surname — at **42 colliding groups over 88 rows** on our 3,117-person corpus, including `N. B. Atkinson` / `N. P. Atkinson`, two people our roster adjudication documents as distinct on five independent grounds. Normalised equality is very likely the right ceiling on your side too, for the same reason.

That residue is tracked at CannObserv/usa-wa#383 — 19 hand-review candidates — held until #501 executes so it does not move the diff under your pass.

**Comment by gregoryfoster** [comment]
## Due diligence: Python name-matching libraries

Before choosing a fold, I checked whether a maintained library already does this. About 20 candidates were screened on PyPI and GitHub activity, licence and dependency weight. The leading ones were then run against this issue's cases plus four pairs that must **not** match. Everything ran in a throwaway venv, with no repo changes.

| pair | want | rapidfuzz `token_set` | Jaro-Winkler | nomenklatura `logic-v2` | nameparser + nicknames rule |
|---|---|---|---|---|---|
| Patterson, curly vs straight quotes | Y | 100 | 100 | 1.00 | ✅ |
| Saling, middle initial | Y | 100 | 93 | 0.79 | ✅ |
| Johnson, middle initial | Y | 100 | 97 | 0.71 | ✅ |
| `Myron “Mike” Kreidler` / `Mike Kreidler` | Y | 76 | 75 | 0.71 | ✅ |
| Kreidler with the old parenthetical | Y | 76 | 75 | **0.16** | ✅ |
| Michael / Mike | Y | 83 | 85 | 0.82 | ✅ |
| `PATTERSON, E. G.` / `E. G. Patterson` | Y | 97 | 66 | 1.00 | ✅ |
| N. B. / N. P. Atkinson (adjudicated distinct, usa-wa) | **n** | **93** | **97** | 0.67 | ✅ rejected |
| Stanley / Samuel Johnson, and two similar pairs | n | 76–90 | 87–92 | 0.50 | ✅ rejected |

**What it shows**

- **String similarity cannot do this job.** The Atkinson pair, two different people, scores above the real Kreidler twin on both metrics, so no threshold separates them.
- **nomenklatura `logic-v2` (OpenSanctions, MIT)** ranks the pairs correctly and explains every score part by part. The margin is thin, though: Atkinson scores 0.67 against 0.71 for the weakest true match. A trailing parenthetical sinks a twin to 0.16 unless it is stripped first. It is also costly:
  - an exact `scikit-learn==1.9.1` pin, plus scipy, numpy, textual and sqlalchemy;
  - PyICU, which has no PyPI wheel, so the VM would need `libicu-dev` and a compiler;
  - about 10 ms per pair.

  That is too heavy for an advisory per-create hint. It would suit a future batch replacement for the admin trigram duplicate detector, alongside splink.
- **nameparser (already a dependency) + `nicknames` (Apache-2.0, zero deps)** give a structural rule: surname equal; given name or quoted nickname equal, an initial of it, or a dictionary nickname; middle initials may be **absent** on one side but never **conflicting**. It scored 11/11 at about 0.2 ms per pair. Caveat: the middle-initial conflict rule was added after Atkinson failed a first draft, so this is the set it was tuned on. Precision on the PM corpus is measured separately (below).
- **The locked nameparser 1.1.3 does not recognise curly-quoted nicknames.** `“Jerry”` lands in `middle`, so quotes must be folded before parsing. nameparser 2.x handles them natively, but it is a new major version with six releases since 2026-07; bumping it is a separate decision.
- **Organisation names:** cleanco's `basename` matched all three legal-form variants tested (`Corp.`, `LLC`, `&`/`and`) and rejected the distinct pairs. It does not strip jurisdiction prefixes, which is acceptable for an equality tier.
- **Ruled out:**
  - probablepeople: `RepeatedLabelError` on Saling; it tags Kreidler's nickname as `MiddleName` and Patterson as a `Corporation`.
  - whoswho: archived in 2019, and raises `IndexError` under nameparser 2.x.
  - namematcher: not on PyPI; last push 2017.
  - hmni: last release 2020; TensorFlow-based.
  - namematch: AGPL.
  - dedupe: needs active-learning labels.
  - DNB `name_matching`: batch TF-IDF.
  - Interzoid: a paid API, so names would leave the box.

## Design going ahead

- **Matching moves into Python.** The store returns the hinted table's raw rows, and one fold/rule is applied to both sides. The Postgres store and the test fake then cannot disagree about what "matches" means.
- **Person tiers**, each labelled on the hint:
  1. `exact`: normalised equality (NFKC, straight quotes, trailing parenthetical dropped, whitespace collapsed, casefold);
  2. `given`: surname plus given name or nickname equal, middle initials compatible;
  3. `initial`: an initial against a given name;
  4. `nickname`: a match through the nicknames dictionary.

  Candidates are blocked on the folded surname.
- **Organisation tiers:** `exact` fold, then `legal_form` (cleanco).
- **Non-public PM names still match**, because a twin may hide there, but such a hint names only the parent and the tier, never the PM-side text. The current query reads `person_names` with no visibility predicate, and the visibility lint cannot see it because the table name is dynamic.
- **Hints remain outside the diff digest**, so this change cannot reset the `--execute` streak.
- **Acceptance fixtures** cover one case per shape: curly quote, middle initial, nickname-as-given, trailing parenthetical and case. Atkinson is a must-not-match.

A collision count of the rule over the live PM corpus follows before merge.
